service/route_handlers.go

```diff
@@ -499,42 +491,69 @@ func getRequestIDAndRedirectParams(r *http.Request, s *model.RancherSamlServiceP
 			continue
 		}
 		claims := token.Claims.(jwt.MapClaims)
-		reqIDs = append(reqIDs, claims["id"].(string))
-		/*
-			HandleSamlLogin sets relayState per request, which contains the URI of the request. This URI has the redirectBackBase and
-			RedirectBackPath values. So we can get the redirectBackBase and redirectBackPath for the current login request from URI of
-			relayState
-		*/
+		rv = append(rv, claims["id"].(string))
+	}
+
+	return rv
+}
+
+func randomBytes(n int) []byte {
+	rv := make([]byte, n)
+	if _, err := saml.RandReader.Read(rv); err != nil {
+		panic(err)
+	}
+	return rv
+}
+
+// HandleSamlAssertion processes/handles the assertion obtained by the POST to /saml/acs from IdP
+func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml.Assertion, serviceProvider *model.RancherSamlServiceProvider) {
+	var redirectBackBaseValue, redirectBackPathValue string
+	secretBlock := x509.MarshalPKCS1PrivateKey(serviceProvider.ServiceProvider.Key)
+
+	if relayState := r.Form.Get("RelayState"); relayState != "" {
+		stateValue := serviceProvider.ClientState.GetState(r, relayState)
+		if stateValue == "" {
+			log.Debugf("cannot find corresponding state: %s", relayState)
+			http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+			return
+		}
+
+		jwtParser := jwt.Parser{
+			ValidMethods: []string{jwt.SigningMethodHS256.Name},
+		}
+		state, err := jwtParser.Parse(stateValue, func(t *jwt.Token) (interface{}, error) {
+			return secretBlock, nil
+		})
+		if err != nil || !state.Valid {
+			log.Debugf("Cannot decode state JWT: %s (%s)", err, stateValue)
+			http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+			return
+		}
+		claims := state.Claims.(jwt.MapClaims)
+		if claims == nil {
+			http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+			return
+		}
 		redirectURI := claims["uri"].(string)
+		log.Debugf("RelayState used to get redirect params: %v", relayState)
 		log.Debugf("redirectURI from claims: %v", redirectURI)
 		query, err := url.Parse(redirectURI)
 		if err != nil {
 			log.Errorf("Error in getting query params")
-			return []string{}, "", "", fmt.Errorf("Error in getting query params: %v", err)
+			return
```

Comment:
If we return over here, how would the saml login flow render? should we redirect and return errorcode?